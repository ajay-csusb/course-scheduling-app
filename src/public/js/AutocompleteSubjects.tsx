import { IItemRendererProps } from '@blueprintjs/select';
import React from 'react';
import { MenuItem } from '@blueprintjs/core';
import { ISubject } from './Subject';

export function filterSubjects(query: string, subject: ISubject): boolean {
  const parsedQuery = query.toLowerCase().trim();
  if (subject.name.toLowerCase().trim().indexOf(parsedQuery) >= 0) {
    return true;
  }
  if (subject.abbr.toLowerCase().trim().indexOf(parsedQuery) >= 0) {
    return true;
  }
  return false;
}

function renderSubjects(subject: ISubject, classRenderer: IItemRendererProps): JSX.Element {
  return (
    <MenuItem
      active={false}
      disabled={false}
      label={subject.abbr}
      key={subject.abbr}
      onClick={classRenderer.handleClick}
      text={subject.name}
      className="search-autocomplete-items"
    />
  );
}

export function renderSubject(subject: ISubject): string {
  return `${subject.name.trim()} (${subject.abbr.trim()})`;
}

export const autocompleteSubjectsProps = {
  itemRenderer: renderSubjects,
  itemPredicate: filterSubjects,
  inputValueRenderer: renderSubject,
};
