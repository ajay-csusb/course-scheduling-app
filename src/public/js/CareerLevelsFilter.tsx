import { IClass, ICareerLevels } from './Class';

export function filter(classes: IClass[], careerLevelsOptions: ICareerLevels): IClass[] {
  const result: IClass[] = [];

  if (!atLeastOneOptionSelected(careerLevelsOptions)) {
    return classes;
  }

  classes.forEach(_class => {
    if (careerLevelsOptions.ugrd && _class.degreeType.toLowerCase() === 'ugrd') {
      result.push(_class);
    }
    if (careerLevelsOptions.pbac && _class.degreeType.toLowerCase() === 'pbac') {
      result.push(_class);
    }
    if (careerLevelsOptions.exed && _class.degreeType.toLowerCase() === 'exed') {
      result.push(_class);
    }
  });

  return result;
}

function atLeastOneOptionSelected(careerLevelsOptions: ICareerLevels) {
  return careerLevelsOptions.ugrd || careerLevelsOptions.pbac || careerLevelsOptions.exed;
}
