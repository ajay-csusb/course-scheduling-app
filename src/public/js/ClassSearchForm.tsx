import * as React from 'react';
import { FormGroup, Button, Label, MenuItem, IOptionProps } from '@blueprintjs/core';
import { Suggest } from '@blueprintjs/select';
import { autocompleteSubjectsProps } from './AutocompleteSubjects';
import { SelectListQuarter } from './SelectListQuarter';
import { SelectListCampus } from './SelectListCampus';
import { ControlGroupMeetingTime } from './ControlGroupMeetingTime';
import { ControlGroupMeetingDay } from './ControlGroupMeetingDay';
import { IMeetingDate, ICareerLevels, ICourseLevels } from './Class';
import { autocompleteInstructorsProps } from './AutocompleteInstructors';
import { ISubject } from './Subject';
import { SelectListInstructionMode } from './SelectListInstructionMode';
import { AdvancedFilterFieldset } from './AdvancedFilterFieldset';

export interface IClassSearchFormProps {
  term: IOptionProps[];
  campus: string;
  subjects: ISubject[];
  meetingDate: IMeetingDate;
  instructionMode: string;
  instructors: string[];
  geClassesAttribute: string;
  geClassesAttributes: IOptionProps[];
  isReset?: boolean;
  startTime: Date;
  endTime: Date;
  courseNo: string;
  classNo: string;
  courseAttr: string;
  sessionCode: string;
  currentTermId: string;
  openClasses: boolean;
  careerLevelOptions: ICareerLevels;
  courseLevelsOptions: ICourseLevels;
  onChangeOfTerm: (event: React.FormEvent) => void;
  onChangeOfCampus: (event: React.FormEvent) => void;
  onChangeOfStartTime: (event: any) => void;
  onChangeOfEndTime: (event: any) => void;
  onChangeOfMeetingDate: (event: any) => void;
  onChangeOfSubject: (event: any) => void;
  onChangeOfCourseNo: (event: any) => void;
  onChangeOfInstructionMode: (event: any) => void;
  onChangeOfInstructor: (event: any) => void;
  onChangeOfCourseAttr: (event: any) => void;
  onChangeOfSessionCode: (event: any) => void;
  onChangeOfClassNo: (event: any) => void;
  onChangeOfGeClassesAttribute: (event: any) => void;
  onChangeOfOpenClasses: (event: any) => void;
  onChangeOfCareerLevelOptions: (event: any) => void;
  onChangeOfCourseLevelsOptions: (event: any) => void;
  onKeyDown: (event: React.KeyboardEvent) => void;
  onSubmit: (event: any) => void;
  onReset: (event: any) => void;
}

export class ClassSearchForm extends React.Component<IClassSearchFormProps, {}> {
  constructor(props: IClassSearchFormProps) {
    super(props);
  }
  public render(): JSX.Element {
    const form = this.getForm();
    return <React.Fragment>{form}</React.Fragment>;
  }

  private getForm(): JSX.Element {
    const subjects = this.getSubjectsAutoCompleteComponent();
    const instructors = this.getInstructorsAutoCompleteComponent();
    const courseNumber = this.getCourseNumberComponent();
    const advancedFilter = this.getAdvancedFilterFieldset();
    return (
      <FormGroup>
        <div className="form-grid">
          <div className="row">
            <div className="col-md-8">
              <div className="row">
                <div className="col-6">
                  <div className="form-item">
                    <SelectListQuarter
                      currentTermId={this.props.currentTermId}
                      term={this.props.term}
                      onChangeOfTerm={this.props.onChangeOfTerm}
                    />
                  </div>
                </div>
                <div className="col-6">
                  <div className="form-item">
                    <SelectListCampus campus={this.props.campus} onChangeOfCampus={this.props.onChangeOfCampus} />
                  </div>
                </div>
                <div className="col-6">
                  <div className="form-item">{subjects}</div>
                </div>
                <div className="col-6">
                  <div className="form-item">{courseNumber}</div>
                </div>
                <div className="col-6">
                  <div className="form-item">{instructors}</div>
                </div>
                <div className="col-6">
                  <div className="form-item">
                    <SelectListInstructionMode
                      instructionMode={this.props.instructionMode}
                      onChangeOfInstructionMode={this.props.onChangeOfInstructionMode}
                    />
                  </div>
                </div>
              </div>
            </div>
            <div className="col-md-4">
              <div className="form-item">
                <ControlGroupMeetingDay
                  meetingDate={this.props.meetingDate}
                  onChangeOfMeetingDate={this.props.onChangeOfMeetingDate}
                  onKeyDown={this.props.onKeyDown}
                />
              </div>
              <ControlGroupMeetingTime
                onChangeOfStartTime={this.props.onChangeOfStartTime}
                onChangeOfEndTime={this.props.onChangeOfEndTime}
                startTime={this.props.startTime}
                endTime={this.props.endTime}
              />
            </div>
            <p>
              <strong>*The majority of Fall classes are online, please select “ALL” to see greater results</strong>
            </p>
          </div>
          <div className="row">
            <div className="col-md-12">
              <div className="form-item">{advancedFilter}</div>
            </div>
          </div>
          <div className="row">
            <div className="col-md-12">
              <div className="form-item">
                <div className="form-buttons">
                  <Button
                    text="Submit"
                    onClick={this.props.onSubmit}
                    type="submit"
                    className="btn btn-primary btn-solid"
                  />
                  <Button
                    text="Reset"
                    onClick={this.props.onReset}
                    type="reset"
                    className="btn btn-secondary btn-solid"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
      </FormGroup>
    );
  }

  private getSubjects(): ISubject[] {
    return this.props.subjects;
  }

  private getInstructors(): string[] {
    return this.props.instructors;
  }

  private isReset(): null | undefined {
    return this.props.isReset ? null : undefined;
  }

  private getSubjectsAutoCompleteComponent(): JSX.Element {
    const SuggestSubject = Suggest;
    return (
      <React.Fragment>
        <Label htmlFor="subject">Course Subject</Label>
        <SuggestSubject
          {...autocompleteSubjectsProps}
          items={this.getSubjects()}
          className="search-autocomplete"
          openOnKeyDown={false}
          resetOnClose={true}
          noResults={<MenuItem disabled={true} text="Searching for subject..." />}
          onItemSelect={this.props.onChangeOfSubject}
          selectedItem={this.isReset()}
        />
      </React.Fragment>
    );
  }

  private getInstructorsAutoCompleteComponent(): JSX.Element {
    const SuggestInstructor = Suggest;
    return (
      <React.Fragment>
        <Label htmlFor="instructor">Instructor</Label>
        <SuggestInstructor
          {...autocompleteInstructorsProps}
          items={this.getInstructors()}
          className="search-instructor-autocomplete"
          openOnKeyDown={false}
          resetOnClose={true}
          noResults={<MenuItem disabled={true} text="Searching for instructor..." />}
          onItemSelect={this.props.onChangeOfInstructor}
          selectedItem={this.isReset()}
        />
      </React.Fragment>
    );
  }

  private getCourseNumberComponent(): JSX.Element {
    return (
      <React.Fragment>
        <Label htmlFor="course-number">Course Number</Label>
        <input
          id="course-number"
          className="course-number"
          type="text"
          placeholder="E.g. 211 for quarter or 2110 for semester"
          dir="auto"
          onChange={this.props.onChangeOfCourseNo}
          value={this.props.courseNo}
          onKeyDown={this.props.onKeyDown}
        />
      </React.Fragment>
    );
  }

  private getAdvancedFilterFieldset(): JSX.Element {
    return <AdvancedFilterFieldset {...this.props} />;
  }
}
