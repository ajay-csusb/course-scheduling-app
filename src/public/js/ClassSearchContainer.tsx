import * as React from 'react';
import { ClassSearchForm } from './ClassSearchForm';
import { ClassSearchResults } from './ClassSearchResults';
import { IClass, Class, IMeetingDate } from './Class';
import { ISubject } from './Subject';
import { UserInput } from './UserInput';
import { Intent, IOptionProps, Callout, Spinner } from '@blueprintjs/core';
import * as ClassSearchUtils from './ClassSearchUtils';
import { MeetingTime } from './MeetingTime';
import { Watchdog } from './Watchdog';
import { InstructionMode } from './InstructionMode';
import { FilterClasses } from './FilterClasses';

interface IClassSearchContainerState {
  term: string;
  campus: string;
  subject: ISubject;
  courseNo: string;
  startTime: Date;
  endTime: Date;
  meetingDate: IMeetingDate;
  instructionMode: string;
  instructorName: string;
  geClasses: boolean;
  geClassesAttribute: string;
  isReset: boolean;
  isLoading: boolean;
  beforeSubmit: boolean;
  noClasses: boolean;
  courseAttr: string;
  sessionCode: string;
  classNo: string;
  showErrorMessage: boolean;
  degreeType: string;
  forceReload: boolean;
}
export class ClassSearchContainer extends React.Component<{}, IClassSearchContainerState> {

  private allResults: IClass[];

  private instructors: string[];

  private subjects: ISubject[];

  private term: IOptionProps[];

  private geClassesAttributes: IOptionProps[];

  private currentTermId: string;

  private readonly dropDownUrl = 'https://webdx.csusb.edu/ClassSchedule/v2/getDropDownList';

  private userInput: UserInput;

  private resultsSection: any;

  constructor(props: any) {
    super(props);
    this.state = this.defaultFormValues();
    this.updateTerm = this.updateTerm.bind(this);
    this.updateCampus = this.updateCampus.bind(this);
    this.updateStartTime = this.updateStartTime.bind(this);
    this.updateEndTime = this.updateEndTime.bind(this);
    this.updateMeetingDate = this.updateMeetingDate.bind(this);
    this.updateSubject = this.updateSubject.bind(this);
    this.updateCourseNo = this.updateCourseNo.bind(this);
    this.updateInstructionMode = this.updateInstructionMode.bind(this);
    this.updateInstructorName = this.updateInstructorName.bind(this);
    this.updateLoadingMessage = this.updateLoadingMessage.bind(this);
    this.updateCourseAttr = this.updateCourseAttr.bind(this);
    this.updateSessionCode = this.updateSessionCode.bind(this);
    this.updateClassNo = this.updateClassNo.bind(this);
    this.updateDegreeType = this.updateDegreeType.bind(this);
    this.updateGeClassAttr = this.updateGeClassAttr.bind(this);
    this.instructorsFound = this.instructorsFound.bind(this);
    this.subjectsFound = this.subjectsFound.bind(this);
    this.termFound = this.termFound.bind(this);
    this.processDropDownListData = this.processDropDownListData.bind(this);
    this.errorProcessingData = this.errorProcessingData.bind(this);
    this.onSubmit = this.onSubmit.bind(this);
    this.onReset = this.onReset.bind(this);
    this.classesFound = this.classesFound.bind(this);
    this.classesNotFound = this.classesNotFound.bind(this);
    this.onEnterKeyPress = this.onEnterKeyPress.bind(this);
    this.allResults = [];
    this.instructors = [];
    this.subjects = [];
    this.term = [];
    this.geClassesAttributes = [];
    this.currentTermId = '';
    this.userInput = new UserInput();
    this.resultsSection = React.createRef();
  }

  public render(): JSX.Element {
    const classSearchResultsComponent = this.getClassSearchResultsComponent();
    const classSearchFormComponent = this.getClassSearchFormComponent();
    const errorMessage: JSX.Element = this.displayErrorMessageWhenSubjectIsEmpty();
    return (
      <React.Fragment>
        <div className="form-section">
          <div className="container">
            {this.state.showErrorMessage && errorMessage}
            {classSearchFormComponent}
          </div>
        </div>
        {this.isLoadingClasses() && <Spinner intent={Intent.PRIMARY} size={25} />}
        {((this.didSubmit() && !this.hasNoClasses()) || (this.didSubmit() && !this.isLoadingClasses())) && classSearchResultsComponent}
      </React.Fragment>
    );
  }
  componentDidMount() {
    ClassSearchUtils.fetchData(this.dropDownUrl, this.processDropDownListData, this.errorProcessingData);
  }

  componentDidUpdate() {
    if (this.resetComplete()) {
      this.setState({
        isReset: false,
      });
    }
    if (this.state.forceReload) {
      this.setState({
        forceReload: false,
      });
    }
  }

  private updateTerm(e: any): void {
    this.setState({
      term: e.target.value,
    });
  }

  private updateCampus(e: any): void {
    this.setState({
      campus: e.target.value,
    });
    this.userInput.setCampus(e.target.value);
  }

  private updateMeetingDate(e: any): void {
    const checkBoxValue = e.target.value;
    this.setState({
      meetingDate: {
        mon: this.toggleMon(checkBoxValue),
        tue: this.toggleTue(checkBoxValue),
        wed: this.toggleWed(checkBoxValue),
        thu: this.toggleThu(checkBoxValue),
        fri: this.toggleFri(checkBoxValue),
        sat: this.toggleSat(checkBoxValue),
        sun: this.toggleSun(checkBoxValue),
      },
    });
  }

  private updateSubject(subject: ISubject): void {
    this.setState({
      subject: subject,
    });
  }

  private updateCourseNo(e: any): void {
    this.setState({
      courseNo: e.target.value,
    });
    this.userInput.setCourseNo(e.target.value);
  }

  private updateInstructionMode(e: any): void {
    this.setState({
      instructionMode: e.target.value,
    });
    this.userInput.setInstructionMode(e.target.value);
  }

  private updateInstructorName(instructor: string): void {
    this.setState({
      instructorName: instructor,
    });
  }

  private toggleMon(checkBoxValue: string): boolean {
    if (checkBoxValue === 'mon') {
      return (!this.state.meetingDate.mon);
    }
    return this.state.meetingDate.mon;
  }
  private toggleTue(checkBoxValue: string): boolean {
    if (checkBoxValue === 'tue') {
      return (!this.state.meetingDate.tue);
    }
    return this.state.meetingDate.tue;
  }
  private toggleWed(checkBoxValue: string): boolean {
    if (checkBoxValue === 'wed') {
      return (!this.state.meetingDate.wed);
    }
    return this.state.meetingDate.wed;
  }
  private toggleThu(checkBoxValue: string): boolean {
    if (checkBoxValue === 'thu') {
      return (!this.state.meetingDate.thu);
    }
    return this.state.meetingDate.thu;
  }
  private toggleFri(checkBoxValue: string): boolean {
    if (checkBoxValue === 'fri') {
      return (!this.state.meetingDate.fri);
    }
    return this.state.meetingDate.fri;
  }
  private toggleSat(checkBoxValue: string): boolean {
    if (checkBoxValue === 'sat') {
      return (!this.state.meetingDate.sat);
    }
    return this.state.meetingDate.sat;
  }
  private toggleSun(checkBoxValue: string): boolean {
    if (checkBoxValue === 'sun') {
      return (!this.state.meetingDate.sun);
    }
    return this.state.meetingDate.sun;
  }

  private classesFound(classes: any): void {
    const transformedClass: IClass[] = [];
    this.allResults = [];
    if (classes === null || classes.length === 0) {
      this.setState({
        noClasses: true,
        isLoading: false,
      });
      return;
    }
    if (classes !== null && classes.length !== 0) {
      classes.forEach((_class: any) => {
        transformedClass.push(Class.transformToClass(_class));
      });
    }

    const validClasses = this.filterClasses(transformedClass);
    const sortedClasses = this.sortClasses(validClasses);
    this.allResults = this.processCourseAttributes(sortedClasses);
    this.setState({
      noClasses: false,
      isLoading: false,
    });
    this.resultsSection.current.scrollIntoView();
  }

  private filterClasses(classes: IClass[]): IClass[] {
    const activeClasses = FilterClasses.filterByActiveClasses(classes)
    const filteredClasses = MeetingTime.filter(activeClasses, this.state.startTime, this.state.endTime);
    return InstructionMode.filter(filteredClasses, this.state.instructionMode);
  }

  private sortClasses(classes: IClass[]): IClass[] {
    return ClassSearchUtils.sortClasses(classes);
  }

  private processCourseAttributes(classes: IClass[]): IClass[] {
    return ClassSearchUtils.parseCourseAttributes(classes, this.geClassesAttributes);
  }

  private classesNotFound(_error: any): void {
    const errorMsg = {
      name: Error().name,
      msg: _error.message,
      stack: Error().stack,
    };
    Watchdog.log(errorMsg);
    this.setState({
      noClasses: true,
      isLoading: false,
    });
  }

  private onSubmit(): any {
    if (this.isSubjectEmpty() && this.areOtherFieldsEmpty()) {
      this.setState({
        showErrorMessage: true,
      });
      return this.displayErrorMessageWhenSubjectIsEmpty();
    }
    this.allResults = [];
    this.setState({
      showErrorMessage: false,
      beforeSubmit: false,
      isLoading: true,
    }, () => {
      this.updateAllClasses();
      }
    );
  }

  private onReset(): void {
    this.setState(this.defaultFormValues());
    this.setState({
      isReset: true,
    }, () => {
      this.userInput = new UserInput();
    });
  }

  private defaultFormValues(): IClassSearchContainerState {
    return {
      term: this.currentTermId,
      campus: 'both',
      subject: {
        name: '',
        abbr: '',
      },
      courseNo: '',
      startTime: new Date('1899-01-01T00:00:00'),
      endTime: new Date('1899-01-01T23:00:00'),
      meetingDate: {
        mon: false,
        tue: false,
        wed: false,
        thu: false,
        fri: false,
        sat: false,
        sun: false,
      },
      instructionMode: 'all',
      instructorName: '',
      geClasses: false,
      isReset: false,
      isLoading: false,
      beforeSubmit: true,
      noClasses: false,
      courseAttr: 'all',
      sessionCode: 'all',
      classNo: '',
      showErrorMessage: false,
      degreeType: 'all',
      geClassesAttribute: '',
      forceReload: false,
    };
  }

  private instructorsFound(data: any): void {
    const instructorsArr: any[] = [];
    instructorsArr.push('All');
    if (data !== undefined && data.instructorList !== undefined) {
      data.instructorList.forEach((_instructor: any) => {
        instructorsArr.push(_instructor.name);
      });
      this.instructors = instructorsArr;
      this.setState({ forceReload: true});
    }
  }

  private subjectsFound(data: any): void {
    const subjects: ISubject[] = [];
    const noOption: ISubject = {
      abbr: 'all',
      name: 'All',
    };
    subjects.push(noOption);
    const subjectsArr = data.abbreviationList;
    if (subjectsArr !== undefined) {
      subjectsArr.forEach((_subject: any) => {
        const subject: ISubject = {
          abbr: '',
          name: '',
        };
        subject.abbr = _subject.subject;
        subject.name = _subject.subject_DESC;
        subjects.push(subject);
      });
      this.subjects = subjects;
    }
    this.setState({ forceReload: true});
  }

  private termFound(data: any): void {
    const terms = data.termList;
    const termArr: IOptionProps[] = [];
    terms.forEach((_term: any) => {
      if (this.hasCurrentQuarterFlag(_term)) {
        this.currentTermId = _term.strm;
      }
      termArr.push({
        label: _term.display_STR, value: _term.strm,
      });
    });
    this.term = termArr;
    this.setState({
      // @Todo unit test this.
      term: this.currentTermId,
      forceReload: true,
    });
  }

  private hasCurrentQuarterFlag(quarter: any): boolean {
    return (quarter.displayed_FLAG === 'Y' && quarter.default_FLG === 'Y');
  }

  private updateStartTime(e: Date): void {
    this.setState({
      startTime: e,
    });
    this.userInput.setStartTime(e);
  }

  private updateEndTime(e: Date): void {
    this.setState({
      endTime: e,
    });
    this.userInput.setEndTime(e);
  }

  private resetComplete() {
    return (this.state.isReset && this.state.beforeSubmit && this.state.subject.abbr.length === 0);
  }

  private isSubjectEmpty(): boolean {
    return (this.state.subject.abbr.length === 0);
  }

  private isInstructorEmpty(): boolean {
    return (this.state.instructorName === '');
  }

  private isCourseNumberEmpty(): boolean {
    return (this.state.courseNo === '');
  }

  private isMeetingDayEmpty(): boolean {
    return (!this.state.meetingDate.mon
    && !this.state.meetingDate.tue
    && !this.state.meetingDate.wed
    && !this.state.meetingDate.thu
    && !this.state.meetingDate.fri
    && !this.state.meetingDate.sat
    && !this.state.meetingDate.sun);
  }

  private isClassNoEmpty(): boolean {
    return (this.state.classNo === '');
  }

  private isValidInstructionModeSelected(): boolean {
    return (this.state.instructionMode !== 'all'
      && this.state.instructionMode !== 'p'
      && this.state.instructionMode !== 'ol'
    );
  }

  private isGeClassesAttributeEmpty(): boolean {
    return (this.state.geClassesAttribute === '');
  }

  private areOtherFieldsEmpty(): boolean {
    if (!this.isInstructorEmpty()) {
      return false;
    }
    if (!this.isCourseNumberEmpty()) {
      return false;
    }
    if (!this.isMeetingDayEmpty()) {
      return false;
    }
    if (!this.isClassNoEmpty()) {
      return false;
    }
    if (this.isValidInstructionModeSelected()) {
      return false;
    }
    if (!this.isGeClassesAttributeEmpty()) {
      return false;
    }
    return true;
  }

  private displayErrorMessageWhenSubjectIsEmpty(): JSX.Element {
    return (
      <Callout className="subject-validation" intent={Intent.WARNING}>
        Please select a Course Subject
      </Callout >
    );
  }

  private hasNoClasses(): boolean {
    return this.isPageLoad() || this.state.noClasses;
  }

  private isPageLoad(): boolean {
    return (!this.state.noClasses && this.allResults.length === 0);
  }

  private getClassSearchResultsComponent(): JSX.Element {
    return (
      <React.Fragment>
        <div ref={this.resultsSection} />
        <ClassSearchResults
          classes={this.allResults}
          currentTerm={this.currentTermId}
          onChangeOfLoadingMessage={this.updateLoadingMessage}
        />
      </React.Fragment>
    );
  }

  private getClassSearchFormComponent(): JSX.Element {
    // @Todo subjects, instructors and terms can be abstracted into
    // a new class.
    return (
      <ClassSearchForm
        term={this.term}
        campus={this.state.campus}
        subjects={this.subjects}
        meetingDate={this.state.meetingDate}
        instructionMode={this.state.instructionMode}
        instructors={this.instructors}
        geClassesAttribute={this.state.geClassesAttribute}
        geClassesAttributes={this.geClassesAttributes}
        isReset={this.state.isReset}
        onChangeOfTerm={this.updateTerm}
        onChangeOfCampus={this.updateCampus}
        onChangeOfStartTime={this.updateStartTime}
        onChangeOfEndTime={this.updateEndTime}
        onChangeOfMeetingDate={this.updateMeetingDate}
        onChangeOfSubject={this.updateSubject}
        onChangeOfCourseNo={this.updateCourseNo}
        onChangeOfInstructionMode={this.updateInstructionMode}
        onChangeOfInstructor={this.updateInstructorName}
        onChangeOfCourseAttr={this.updateCourseAttr}
        onChangeOfSessionCode={this.updateSessionCode}
        onChangeOfClassNo={this.updateClassNo}
        onChangeOfGeClassesAttribute={this.updateGeClassAttr}
        onSubmit={this.onSubmit}
        onReset={this.onReset}
        onKeyDown={this.onEnterKeyPress}
        startTime={this.state.startTime}
        endTime={this.state.endTime}
        courseNo={this.state.courseNo}
        classNo={this.state.classNo}
        courseAttr={this.state.courseAttr}
        sessionCode={this.state.sessionCode}
        currentTermId={this.state.term}
      />
    );
  }

  private isLoadingClasses(): boolean {
    return this.state.isLoading;
  }

  private didSubmit(): boolean {
    return !this.state.beforeSubmit;
  }

  private updateLoadingMessage(): void {
    if (this.state.isLoading) {
      this.setState({
        isLoading: false,
      });
    }
  }

  private updateCourseAttr(e: any) {
    const selectedValue = e.target.value;
    // @Todo simplify this.
    this.updateDegreeType('all');
    if (this.isDegreeType(selectedValue)) {
      this.updateDegreeType(selectedValue);
      return;
    }
    this.setState({
      courseAttr: selectedValue,
    });
    this.userInput.setCourseAttr(selectedValue);
  }

  private updateSessionCode(e: any) {
    this.setState({
      sessionCode: e.target.value,
    });
    this.userInput.setSessionCode(e.target.value);
  }

  private updateClassNo(e: any) {
    this.setState({
      classNo: e.target.value,
    });
    this.userInput.setClassNo(e.target.value);
  }

  private updateDegreeType(value: string) {
    this.setState({
      degreeType: value,
    });
    this.userInput.setDegreeType(value);
  }

  private updateGeClassAttr(e: any) {
    this.setState({
      geClassesAttribute: e.target.value,
    });
    this.userInput.setGeClassesAttr(e.target.value);
  }

  private processDropDownListData(data: any): void {
    this.termFound(data);
    this.instructorsFound(data);
    this.subjectsFound(data);
    this.geClassesAttributesFound(data);
  }

  private errorProcessingData(_error: any): void {
    const errorMsg = {
      name: Error().name,
      msg: _error.message,
      stack: Error().stack,
    };
    Watchdog.log(errorMsg);
  }

  private updateAllClasses() {
    this.userInput.setTerm(this.state.term);
    this.userInput.setInstructor(this.state.instructorName);
    this.userInput.setSubject(this.state.subject.abbr);
    this.userInput.setMeetingDay(this.state.meetingDate);
    Class.getAllClasses(this.classesFound, this.classesNotFound, this.userInput);
  }

  private onEnterKeyPress(event: React.KeyboardEvent): void {
    if (event.key === 'Enter') {
      this.onSubmit();
    }
  }

  private isDegreeType(degreeAbbr: string): boolean {
    return (degreeAbbr === 'UGRD' || degreeAbbr === 'PBAC' || degreeAbbr === 'EXED');
  }

  private geClassesAttributesFound(data: any): void {
    const geClasses = data.classAttributeList;
    const noOption: IOptionProps = {
      value: '',
      label: 'All',
    };
    this.geClassesAttributes.push(noOption);
    geClasses.forEach((attribute: any) => {
      if (attribute.crse_ATTR === 'GE') {
        this.geClassesAttributes.push({
          label: attribute.descr,
          value: attribute.descr.split(' ')[0],
        });
      }
    });
    this.setState({
      forceReload: true,
    });
  }

}
