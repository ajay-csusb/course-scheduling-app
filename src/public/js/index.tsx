import * as React from 'react';
import * as ReactDOM from 'react-dom';
import { ClassSearchContainer } from './ClassSearchContainer';
import { Utils } from './Utils';

Utils.loadExternalTools();
if (Utils.isProd()) {
  const ErrorBoundary = Utils.loadBugsnagComponent();
  ReactDOM.render(<ErrorBoundary><ClassSearchContainer /></ErrorBoundary>,
    document.querySelector('#class-search-form-wrapper'));
} else {
  ReactDOM.render(<ClassSearchContainer />, document.querySelector('#class-search-form-wrapper'));
}
