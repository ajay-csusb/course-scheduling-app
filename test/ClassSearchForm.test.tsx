import { mount } from 'enzyme';
import * as React from 'react';
import { ClassSearchContainer } from '../src/public/js/ClassSearchContainer';
import { TestUtils } from './TestUtils';
import { ClassSearchForm } from '../src/public/js/ClassSearchForm';
// tslint:disable:max-line-length

describe('Class search form', () => {
  beforeAll(() => {
    TestUtils.ajax();
  });

  describe('When a search is performed without entering subject', () => {
    let classSearchContainerWrapper = null;
    beforeEach(() => {
      classSearchContainerWrapper = mount(<ClassSearchContainer />);
      classSearchContainerWrapper.find('button[type="submit"]').simulate('click');
    });

    it('should display error message', () => {
      expect(classSearchContainerWrapper.html()).toContain('Please select a Course Subject');
    });
  });

  describe('When a user selects a GE course attribute', () => {
    let searchFormComponent = null;
    describe('and does not select a subject', () => {
      beforeEach(() => {
        searchFormComponent = mount(<ClassSearchContainer />);
        searchFormComponent.find('#additional-filters').simulate('click');
        searchFormComponent.find('.select-ge-classes-attr select').simulate('change', { target: { value: 'foo' } });
        searchFormComponent.find('button[type="submit"]').simulate('click');
      });

      it('should display error message', () => {
        expect(searchFormComponent.html()).not.toContain('Please select a Course Subject');
      });
    });
  });

  describe('When a user performs a search', () => {
    let searchFormComponent = null;
    let classSearchContainerComponent = null;

    beforeEach(() => {
      classSearchContainerComponent = mount(<ClassSearchContainer />);
      searchFormComponent = classSearchContainerComponent.find(ClassSearchForm);
    });

    it('should pass the correct props to class search form component', () => {
      expect(searchFormComponent.prop('openClasses')).toBeFalsy();
      expect(searchFormComponent.prop('careerLevelOptions')).toEqual({ugrd: false, pbac: false, exed: false});
      classSearchContainerComponent.find('#additional-filters').simulate('click');
      classSearchContainerComponent.find('.open-classes > input').simulate('change');
      classSearchContainerComponent.find('input#ugrd').simulate('change');
      classSearchContainerComponent.find('input#pbac').simulate('change');
      classSearchContainerComponent.find('input#exed').simulate('change');
      classSearchContainerComponent.find('button[type="submit"]').simulate('click');
      searchFormComponent = classSearchContainerComponent.find(ClassSearchForm);
      expect(searchFormComponent.prop('openClasses')).toBeTruthy();
      expect(searchFormComponent.prop('careerLevelOptions')).toEqual({ugrd: true, pbac: true, exed: true});
    });
  });
});
