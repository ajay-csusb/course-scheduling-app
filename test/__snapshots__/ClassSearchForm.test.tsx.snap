// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test class search form component Class search form component snapshot 1`] = `
<div>
  <Blueprint3.FormGroup
    label="Fill in one or more of the fields below:"
  >
    <Blueprint3.RadioGroup
      className="bp3-inline quarter-select"
      inline={true}
      label="Quarter"
      onChange={[Function]}
      selectedValue="current"
    >
      <Blueprint3.Radio
        label="Winter 2019"
        value="current"
      />
      <Blueprint3.Radio
        label="Spring 2019"
        value="next"
      />
    </Blueprint3.RadioGroup>
    <Blueprint3.RadioGroup
      inline={true}
      label="Campus"
      onChange={[Function]}
      selectedValue="both"
    >
      <Blueprint3.Radio
        label="Both"
        value="both"
      />
      <Blueprint3.Radio
        label="San Bernardino"
        value="san-bernardino"
      />
      <Blueprint3.Radio
        label="Palm Desert"
        value="palm-desert"
      />
    </Blueprint3.RadioGroup>
    <div
      className="bp3-input-group bp3-inline"
    >
      <Component
        className="bp3-inline"
        htmlFor="course-number"
      >
        Course Number
      </Component>
      <input
        className="bp3-input bp3-inline"
        id="course-number"
        placeholder="begins with"
        type="text"
      />
    </div>
    <Blueprint3.RadioGroup
      inline={true}
      label="Desired Meeting Time"
      onChange={[Function]}
      selectedValue="all"
    >
      <Blueprint3.Radio
        label="All"
        value="all"
      />
      <Blueprint3.Radio
        label="Before Noon"
        value="before-noon"
      />
      <Blueprint3.Radio
        label="After Noon"
        value="after-noon"
      />
      <Blueprint3.Radio
        label="4 PM and later"
        value="4pm-later"
      />
    </Blueprint3.RadioGroup>
    <div
      className="bp3-select"
    >
      <select
        onChange={[Function]}
      >
        <option
          selected={true}
          value="all"
        >
          Meeting Date
        </option>
        <option
          value="mon"
        >
          Monday
        </option>
        <option
          value="tue"
        >
          Tuesday
        </option>
        <option
          value="wed"
        >
          Wednesday
        </option>
        <option
          value="thu"
        >
          Thursday
        </option>
        <option
          value="fri"
        >
          Friday
        </option>
        <option
          value="Sat"
        >
          Saturday
        </option>
        <option
          value="Sunday"
        >
          Sunday
        </option>
      </select>
    </div>
    <hr />
    <Blueprint3.Button
      className="bp3-intent-primary"
      large={true}
      text="Submit"
      type="submit"
    />
    <Blueprint3.Button
      large={true}
      text="Reset"
      type="reset"
    />
  </Blueprint3.FormGroup>
</div>
`;
